datumaro/components/importer.py

```diff
@@ -177,3 +180,70 @@ def _find_sources_recursive(
                 if sources:
                     break
         return sources
+
+
+def with_subset_dirs(input_cls: Importer):
+    @wraps(input_cls, updated=())
+    class WrappedImporter(input_cls):
+        NAME = input_cls.NAME
+
+        @classmethod
+        def detect(
+            cls,
+            context: FormatDetectionContext,
+        ) -> Optional[FormatDetectionConfidence]:
+            @contextmanager
+            def _change_context_root_path(context: FormatDetectionContext, path: str):
+                tmp = context.root_path
+                context._root_path = path
+                yield
+                context._root_path = tmp
+
+            confs = []
+            path = context.root_path
+
+            if not osp.isdir(path):
+                context.fail(
+                    f"{input_cls.NAME} should require an input as a directory path. "
+                    f"However, {path} is not a directory path."
+                )
+
+            for sub_dir in os.listdir(path):
+                sub_path = osp.join(path, sub_dir)
+                if osp.isdir(sub_path):
+                    with _change_context_root_path(context, sub_path):
+                        conf = input_cls.detect(context)
+                    if conf is not None:
+                        confs += [conf]
+
+            if len(confs) == 0:
+                context.fail(f"{input_cls.NAME} cannot find its subdirectory structure.")
+
+            return max(confs)
+
+        def __call__(self, path, **extra_params):
+            sources = []
+            for sub_dir in os.listdir(path):
+                sub_path = osp.join(path, sub_dir)
+                if osp.isdir(sub_path):
+                    source = input_cls.__call__(self, sub_path, **extra_params)
+
+                    if len(source) != 1:
+                        raise DatasetImportError(
+                            f"@with_subset_dirs only allows one source format from {sub_path}."
+                        )
+
+                    if "subset" in source[0]:
```

Comment:
In this case, the imported dataset with the `with_subset_dirs` decorator should only be `subset` rather than `dataset`, is this correct?

Reply:
No, the imported dataset will ultimately have 
```python
subsets: {
    sub_dir: DatasetBase(subset=sub_dir)
    for sub_dir in os.listdir(path)
}
```
where `DatasetBase` type corresponds to the importer's `DatasetBase` type.